### DISCUSSION TITLE
Can anything be done about Crackling/Static ocassionally when playing?

### DISCUSSION THREAD

**Discussion by MDoder97** [CoderLine/alphaTab#558] [Q&A] [answered]
### Expected Results
No crackling.

### Observed Results
Crackling every 10-30 seconds on most tabs

### Steps to Reproduce
Playing most tabs will produce this.
It seems to be a rendering / lag issue. If you point me in the right direction I'll have a look at the source too.

Also I'd be thankful if you could feel me in on this question -> https://github.com/CoderLine/alphaTab/discussions/553

Thanks.

**Comment by Danielku15** [comment]
I cannot confirm that there is a general bug in alphaTab which would cause this. To me it rather sounds your specific setup is suffering of performance issues (your device or your page). I recommend opening a [Discussion](https://github.com/CoderLine/alphaTab/discussions) where you provide more insight on your precise setup and tests you have done.

**Comment by MDoder97** [comment]
I should have been more detailed.
This actually happens on the default implementation you provided as well.
Try this tab for example in the default setup - https://firebasestorage.googleapis.com/v0/b/guitarflow-95c71.appspot.com/o/tabs%2FRed%20Hot%20Chili%20Peppers%20-%20Californication.gp3?alt=media&token=21ce9eb8-eee8-4173-a595-a9a5f861f3da

It should happen somewhere in the first 2-3 lines
Try this and let me know if you hear it.

**Reply by Danielku15** [reply]
I tried it and could not hear or see any significant issues, not in the default playground samples, not if I drag it to the player on the alphatab.net website. 🤔  Maybe you can do a performance profile on your machine to see if there is something spiking or overloaded? Most browsers have a profiler in their developer tools you can use as a start. This might lead us a bit further. 

Sound issues usually mean that the PC cannot handle the load to synthesize the audio, so it starts to buffer and continues to playing then, this can cause a cracking noise. I have tested alphaTab on a wide range of mobile devices via Browserstack but did not notice any issues in the past. Maybe you can share some hardware specs you are running on?

**Reply by MDoder97** [reply]
@Danielku15 After testing on some devices, It does appear to be a hardware issue with worse specs resulting in more crackling. My laptop has an old i5 with integrated graphics, my pc is beefy so no crackling there.

Slightly more crackling on my xiaomi redmi note 7 phone and a lot of crackling on old Samsung S2.

My CPU is hovering around 20% while playing on laptop so it doesn't seem to be overloaded. I find it strange that my xiaomi phone is also crackling, the specs should be enough.

Can anything be done about this? Perhaps some sort of buffering in advance?

**Reply by Danielku15** [reply]
As said: please do a profiling on your affected environment. Then we will know for sure what is the causing it. 
alphaTab is buffering already, but there is rather a tendency that the cursor animations eat up the CPU. You could try [disabling the cursors ](https://alphatab.net/docs/reference/settings/player/enablecursor#description) as a test. 

I already have https://github.com/CoderLine/alphaTab/issues/480 in my backlog which might improve the performance because the full audio parts would be in a worker. But AudioWorklets require HTTPS to work. I think I will never get who adds such restrictions to specs. I have access to WebAudio and WebWorkers without HTTPs, but using WebAudio in WebWorkers requires HTTPS? 🤦‍♂️

**Reply by MDoder97** [reply]
@Danielku15 I started debugging this but as I said, even the copied barebones implementation you provided cause this so I'm not sure how to do profiling here. Similar static can be heard sometimes when looping certain parts the moment cursor snaps back. I can send you a video from different devices I'm using so you can see what I'm talking about if you want. 
I'm in if you need a hand to figure it out.

Also, I didn't want to open another discussion to ask this but, do you you have any resource for good electric guitar soundfonts viable for alphaTab? As in, not 20mb in size but more like 1.5-2max? I've been scouring the net but the only good ones are also really large, the one you provided turns out to be the most versatile.

**Reply by Danielku15** [reply]
Videos will not allow me to analyze the problem. I recommend you do some research on how to do javascript profiling and then do some tests on your environment. Here some links to get started: 

https://developers.google.com/web/tools/chrome-devtools/rendering-tools
https://developer.mozilla.org/en-US/docs/Tools/Performance
https://blog.appsignal.com/2020/02/20/effective-profiling-in-google-chrome.html

> Also, I didn't want to open another discussion to ask this but

Please exactly do that. Keep one discussion about one topic.

**Reply by Danielku15** [reply]
@MDoder97 Were you able to find out anything more in your tests? I would be eager to find out what is causing this performance issues for you. I did some research but without narrowing down the root cause any improvements will be hard to do. 

There might be tricks like [described here](https://stackoverflow.com/questions/13176746/css-keyframe-animation-cpu-usage-is-high-should-it-be-this-way). If the cursor animations are burning too much CPU, a CSS style might move them to the GPU. 
```css
.at-cursors {
    transform: translateZ(0);
}
```
Or we can make a feature to disable animations and only have a "ticking" cursor. We could also try to detect high CPU usage and try to make things less CPU intense (e.g. disable or reduce animation frames, switch non-smooth scrolling). 

The `will-change` property might boost as well some things like: 

```css
.at-cursor-bar, .at-cursor-beat {
    will-change: left, top;
}
```

Or moving from a left/top to a `translate` approach might also be faster. 

But these are all proposals related to the scrolling and animations. If the actual audio synthesis is to blame. Maybe a change of buffer sizes might do the trick.  

Experience-wise strange audio noises are coming from the fact that the main UI thread is overloaded (from other things) and then the Web Audio API behaves weird. Audio Worklets might solve this for modern browsers on HTTPs.

**Reply by Danielku15** [reply]
Ping @MDoder97 : You got any more insights already on this topic?

**Reply by MDoder97** [reply]
@Danielku15 Hey Daniel, sorry for not responding. I've been completely swamped with work lately so I haven't been able to test it yet. I'll try out everything you mentioned here and let you know, hopefully in the next week or so.

**Comment by Reiji777** [comment]
@Danielku15 Not sure if this is what exactly @MDoder97 is experiencing: on my end when accessing the tutorial on the official website, the tab can't be played smoothly on one of my Android devices, meaning I could hear the glitches here and there, and also the tab is not played at a consistent BPM. On iOS and PC it works great.

The spec of my Android device is not very bad I assume:
- Model: Samsung SM-A750C
- CPU: Exynos 7885 (2.2GHz Dual-core (Cortex®-A73) +1.6GHz Hexa-core (Cortex®-A53))
- Android version: 10
- Browser: Chrome 91.0.4472.88

Weirdly, I have an older Android device on which the playback works fine.
- CPU: 4 x 2.1 GHz + 4 x 1.7 GHz
- Android version: 8
- Browser: Chorme 90.0.4430.210

I also try to turn off the cursor completely by setting `enableCursor` to be `false`,, but still the issue exists. 
The code is here (identical to the official tutorial except for the disabled cursor): https://jsbin.com/solatavuke/

I did a basic profiling but can't see anything that's causing the poor performance. Please see below the profiling screenshot. I did it through `chrome://inspect` on my PC connecting the Android device.

![image](https://user-images.githubusercontent.com/41455097/121796415-c090d080-cc53-11eb-8d93-c86e7bbf2a09.png)

Though maybe not helpful, here's a short video to demonstrate the issue. 
https://drive.google.com/file/d/1AS9Y2242YzEe-L6PaaHAm1wj4VDksB0k/view?usp=sharing

Let me know if you need any further info to hunt down the issue. Happy to provide more info.

Also do you still plan to work on #480? I'm not very familiar with the Web Audio API, but if there's anything I can contribute or sponsor please let me know. Thanks again for your efforts on this project!

**Reply by Danielku15** [reply]
Thanks for the deeper insights but it's still quite hard to tell what's happening here. Can you save this profiling snapshot and share it here? When you right click in chrome you can save the profile as JSON.

**Reply by Reiji777** [reply]
@Danielku15 Here's the profiling snapshot. Seems Github doesn't support json as an attachment so I zipped it. Let me know if you need anything else. 
[Profile-alphatab-20210613T211532.zip](https://github.com/CoderLine/alphaTab/files/6643855/Profile-alphatab-20210613T211532.zip)

**Reply by Reiji777** [reply]
Hey @Danielku15 just wanna make sure if the profiling snapshot above looks good enough, or do you need any further info for the investigation?

**Reply by Danielku15** [reply]
@Reiji777  The profiling snapshot is fine and I can load it. Unfortunately I currently have not much time for alphaTab as my work and private life is quite busy. I had a look and in general alphaTab is performing quite good regarding the time it needs to generate audio samples. There is quite some time between the intervals where audio is again generated on the background worker. If it is easy to reproduce for you, can you maybe try to collect a test dataset for me consisting of the following setup: 

* Setup information:
  * The exact phone model 
  * The exact android version on your phone
  * The exact browser version installed
* A test run with
  * A separate video/audio recording showing the crackling
  * A chrome profile via remote debugger
  * A rough description at which seconds in the timeline you experienced the crackling. 

Goal is to align the actual crackling with the profiling timeline. 

**Technical insights:**
I see some suspicious patterns I would not expect but I will have to make some further tests and dig a bit deeper here. One example is that normally alphaTab buffers up to 10 times a buffer size, and if the buffer size falls under 1/2 of the total buffer size, it requests this missing half. This way alphaTab ensures that the buffer is always half full while playing and has some "reserves" in case the system stutters a bit. But in the profile I can see that on every generation it generates full 10 buffers which would mean that the actual audio buffer is empty and we could have the stuttering. ​

Once I have some further insights, I might provide you guys a more dedicated example to test on your phones. I will need to add some particular logs related to buffer sizes, sampling rates, request intervals etc. I fear that the browser on these systems might not respect certain settings on the audio buffer. This might have to do with the underlying audio driver on the system, because Chrome itself is quite deterministic regarding the Web Audio implementation. 

Will try to keep you up to date.

**Reply by Danielku15** [reply]
@Reiji777 @lannka I quickly patched something together and added some basic changes. Can you repeat your tests on this page and provide me the full console log trace of the browser: 

https://demo.alphatab.net/crm/crackling/#

alphaTab now buffers 5 seconds and if the buffer hits 2.5sec of remaining audio, it again generates 2.5sec of new audio data. The logs contain a lot of details about buffer sizes, sampling rates etc. I hope this gives a good indication why you might experience crackling.

**Reply by Reiji777** [reply]
@Danielku15 Appreciate your thoughts! I did a test and here's the video, profiling snapshot and the full console log of my trying the demo page above with my Android device. I couldn't specify when I was experiencing the cracklings as it was basically everywhere as you can see in the video. Let me know if you need any further info.

https://drive.google.com/file/d/1oWKPc9PmIAtW6m3AB9k2aG6jvSgFy99n/view?usp=sharing

- Model: Samsung SM-A750C
- CPU: Exynos 7885 (2.2GHz Dual-core (Cortex®-A73) +1.6GHz Hexa-core (Cortex®-A53))
- OS: Android 10
- Build version: QP1A.190711.020.A750CONU1BUE1
https://network.mobile.rakuten.co.jp/information/software/galaxy-a7/
- Browser: Chrome 91.0.4472.120


On a side note, I tried to double the buffersizes in `AlphaSynthWebAudioOutput` (the 3 places in the image below, and `AlphaSynthWebAudioOutput.BufferSize`) to 8192 in alphatab 1.2.1, and on my Android device the crackling was just gone. Doubling it again to 16384 would cause the cracklings though. I'm not sure if this makes any sense but just to share with you.

![image](https://user-images.githubusercontent.com/41455097/123631685-fdfb7d80-d851-11eb-9f01-fb43b7ebe19d.png)

**Reply by Danielku15** [reply]
I found something suspicious on the sampling rate settings that could cause some weird things. Increasing the buffer size is not the best choice. This small buffer size ensure that the signaling on the changed time is fine-grained enough to catch all relevant updates. 

I uploaded a new version with the cursor disabled and one change related to the sampling rate handling and buffer size calculation. Can you give it again a test (be sure to refresh until your browser cache updates). The new version is indicated by the fact that the cursor is hidden.

**Reply by Reiji777** [reply]
@Danielku15 Here's the test result of the latest demo without the cursor. The cracklings are improved comparing to my previous test, while I could still hear it time to time, and I notice that some notes are skipped without being played, which you could find at the following seconds in my video:  0:07, 0:09, 0:15, 0:10, 0:23, 0:27

https://drive.google.com/file/d/1ZQDOJ_0B09Ba3wL5WasBZwb9DOSPea1S/view?usp=sharing

**Reply by Danielku15** [reply]
I think I have found out what causes the issue. I found some very suspicous logs which definitely explain the crackling and behaviors you might see. I entered the following bug to track the fix: 
https://github.com/CoderLine/alphaTab/issues/631

Thanks a lot for providing these logs. Will try to provide a fix soon.

**Reply by Danielku15** [reply]
Update: I was able to cause the issue locally with some artificial delays and worked out a fix. (PR open at https://github.com/CoderLine/alphaTab/pull/632)
Can you try again on [the same link](https://demo.alphatab.net/crm/crackling/) whether the fix works out for your phones? To ensure to have a realistic load I re-enabled the cursor. Once I get the confirmation it is fixed for you I'll merge the PR. 

Ping @Reiji777  @lannka  @MDoder97

**Reply by Reiji777** [reply]
@Danielku15 Thanks for the update! As I was testing on my Android device, the crackling is much better than before, but still noticeable time to time (e.g. 00:07, 00:12, 00:15 )

Seems the debug logs have been removed so I didn't include the log file. Let me know if you need any further info.

https://drive.google.com/file/d/12RyWHn87Hh6RLgEYX5TXsmVolbyw6Dpp/view?usp=sharing

**Reply by lannka** [reply]
@Danielku15 I tested the link on my phone, I think you fixed it. Didn't notice a single crackling throughout the whole song.

**Reply by Danielku15** [reply]
@Reiji777 I had a look at the profile, and it seems the browser seems quite overloaded due to some cursor and element highlighting actions. It causes a rather expensive layouting in the browser. As a temporary workaround I could add some settings which allows disabling some actions like the highlighting and the animated cursor. Via some custom browser/device detection you could enable/disable these features accordingly. 

It might take a bit longer to find a new approach for the cursor which performs better.

**Reply by Reiji777** [reply]
@Danielku15 Thanks for investigation! The additional settings would be super appreciated! I will try it out by disabling some of the animations when the settings are available.

**Reply by Danielku15** [reply]
@Reiji777  @lannka  Can you try again on the same test url whether the situation improved? I adapted some bits on the cursor positioning. I still have some more ideas but would like to check if the changes already make a difference.

**Reply by lannka** [reply]
thanks Daniel. I tried again on my phone. Performance wise I didn't notice a change, as it already worked smoothly on my phone with your previous buffering improvement.

But I did notice a new regression with your cursor change , it now animates unexpectedly when moving to the next line.

**Reply by Reiji777** [reply]
@Danielku15 On my end I do feel that the cracklings are improved comparing to my previous video. (links at the bottom)
There are still some minor cracklings noticeable, and I notice that the cursor animation is not as smooth as previously.

Here're the quick video records for your reference:
Latest: https://drive.google.com/file/d/1IpfNRna8kD-f0B33_rsYfiMEGhD2QpVv/view?usp=sharing
Previous: https://drive.google.com/file/d/12RyWHn87Hh6RLgEYX5TXsmVolbyw6Dpp/view?usp=sharing

**Reply by Danielku15** [reply]
@Reiji777 The fact that the animation is not smooth might be the reason for the crackling. Your phone likely is lacking of the required power to handle the UI animations. Can you also provide me the profiling results like last time? Would be interesting to get some insights on the performance metrics. On my desktop profile I could still see some interesting call stacks which are triggered regularly, but on my PC they are just taking microseconds. 

I need to decide on whether it is really worth changing from left/top to a transition-translate approach. As @lannka has indicated, it will require some more adaption to avoid a regression bug on flickering cursors.  

The tricky part is: I have various ideas on how I could reduce the load, but it will come at some cost of features. It would be great if I can detect somehow if the browser is overloaded. 

Here my thoughts: 
- The synthesizer provides quite detailed when the playback position changes. Every position change, triggers some logic to position and animate the cursor if required. This should be usually fast because the lookup of the currently played beat should be very fast. If there is a change, we currently place the element via left/top which seem to trigger a re-layout. 
- I optimized now the code to use not left/top but a transition-translate which should not trigger this re-layouting.
- But there is still a change of width/height involved on line changes or bar changes (as bars have different widths/heights). This I could also try to change to a system not triggering a layouting in the browser. Maybe with using a transition-scale and some custom checks.
- I could try to reduce frequency of cursor updates. There are many ways to improve this area too.  The resolution might be higher than we actually need it or I could even try to add some custom stuff into the synthesizer which signals beat changes. 
- As mentioned I could also disable the animations on the beat cursor. A setting would be easy, but actually it would be a pity to disable it always because there might be poorly performing devices. Hence I would like to detect the performance in some way and disable the beat animation dynamically. 
- jQuery also causes quite some CPU cost when triggering events. Removing jQuery on your page if you don't need it might also improve the performance a bit. 

As a general not-so-technical feedback: After my initial fixes, the crackling is just a side-effect of the device being overloaded and not having enough CPU resources. Hence I am trying to find out how to reduce the CPU load to eliminate the crackling.

**Reply by Reiji777** [reply]
@Danielku15 Thanks for sharing your thoughts in detail! Here's the profiling result with a video (not much different from the last one). Hope this could be helpful.

https://drive.google.com/file/d/12iglgHh1urfkpIj9zoLUJsrlTjdWlq9R/view?usp=sharing

**Reply by lannka** [reply]
Thanks @Danielku15 for the details.

**Avoid relayout**
I feel this is promising. As you pointed out, this might be completely avoided with extra logic and transition/scale.

**Reduce frequency**
It seems to me the current setup is already optimal onPositionChange (I assume this is triggered on beat change) + requestAnimationFrame. What's your new idea?

**Load detection**
There is [LongTaskObserver](https://developer.mozilla.org/en-US/docs/Web/API/Long_Tasks_API) but I'm not sure if it's suitable for the case here. Also it lacks browser support.


I think your initial fixes already fixed most cases. I tried a couple of my test phones, they are all quite good now. Agree that the remaining is mostly performance optimization.

**Reply by Danielku15** [reply]
@Reiji777 I think we might be getting closer. I added now again some more micro optimizations:
1. I try to minimize some layout intense actions during cursor updates. In my tests the number of costly layout tree updates was improved quite significant and also the dropped frames due to long operations reduced. 
2. I also added a new option which disables triggering alphaTab events through the DOM and jQuery. This requires the developer then to hook up the events via API object directly, but saves some milliseconds because we do not spam out all events through 3 channels

It would be awesome if you can run a profiling session again on your device.

**Reply by Reiji777** [reply]
@Danielku15 Thanks for the update! As I was testing and profiling on the demo page, I notice it is indeed improved comparing to last time. There's some minor crackling but overall much smoother. 

Below is the profiling result for your info:
https://drive.google.com/file/d/1GrLZWZLqTv6xmfdYy2OulDca9nUp4jnS/view?usp=sharing

**Reply by Danielku15** [reply]
@Reiji777 And thanks to you for all the profiling sessions. While I can test some parts with the CPU throttling feature, this phone really struggles with some operations which usually are really fast. 

Here some findings from your latest session: 
- It's not only alphaTab putting pressure to the phone but also other operations like fontawesome checking for new icons has some really unusual long timeline entries. 
- Whenever the lazy loading kicks in, some additional pressure comes to the phone due to the SVG parsing of the next row. 
- There was one last performance critical operation I could identify and have improved.

I created now a new version which also allows setting some options. You can disable the animated cursor, and the individual element highlighting via query parameters. I personally think the phone hardware suffers of some hardware acceleration which makes some actions unusually slow. It would be great if you can profile the following combinations. I suspect the highlighting to cause expensive layout updates. To improve this, likely a full redesign of the highlighting feature would be needed. It currently is based on CSS classes which trigger a style recalculation and an update of the layer tree. 

**All features active**
https://demo.alphatab.net/crm/crackling/?animate=1&highlight=1
**Animated cursor but no highlighting**
https://demo.alphatab.net/crm/crackling/?animate=1&highlight=0
**Not animated but highlighting**
https://demo.alphatab.net/crm/crackling/?animate=0&highlight=1
**Not animated and no highlighting**
https://demo.alphatab.net/crm/crackling/?animate=0&highlight=0

Ultimately I would say the AudioWorklets are needed to ultimately get rid of the crackling. The UI thread seems quite under pressure on this phone. But even with AudioWorklets, the UI might become a bit laggy unless we manage to get rid of these rendering related (violet) timeline entries. 

Let's see how the feature flags and the latest changes perform.

**Reply by Danielku15** [reply]
Update 2: I just gave it a very brief try to use Audio Worklets if available in the browser. It is just a very early attempt which follows the same approach of the old architecture to synthesize the audio in a WebWorker, and the put it to a ringbuffer within the worklet. 

This way the Audio Output can pull the samples from the worklet without putting pressure to the main UI thread. There is still a lot of communication going via UI thread to pass on various messages. I see some chances to maybe better off-load stuff from the UI but it's a start worth investigating on performance side.  

It is based on the branch with all other improvements and allows the same parameters as described above: https://demo.alphatab.net/crm/audioworklets/

**Reply by Reiji777** [reply]
@Danielku15 I profiled all the four variations plus the audioworklet one. Despite the bugs (weird sounds at the beginning — seems several notes are played at the same time; cursor runs later than the actual note being played), the audioworklet version runs smoothly without cracklings. 

https://drive.google.com/file/d/1OqiJdzlKeYhLktPRc9ruWw6aKeKjojP6/view?usp=sharing

I also tested the audioworklet version on an old Android device with a pretty weak spec ([fire hd 8 6th generation](https://gadgets.ndtv.com/amazon-fire-hd-8-2016-3769)), and it runs smoothly even on that.

Also I noticed a bug for all the versions above: if I click on a note to move the cursor when playback is stopped, and I start playing again, the cursor won't move to show the playback status any more.

**Reply by Danielku15** [reply]
The current improvements are quite patchy and not well designed. I will definitly need some hours to get them clean and regression tested. After we were not really improving the situation I rather went the straight forward approaches to get some improvements in to profile them again. 

Unfortunately the improvements on the highlighting and animation did not really bring the desired effect on this device. It's really crazy how these slow "update layer trees" won't disappear. My own code really performs well and fast. But there seem to be side effects in the engine causing real trouble. 

I will try to prepare some proper version soon with the audio worklets and clean improvements in place.

**Reply by Danielku15** [reply]
@Reiji777 I uploaded a new version of the audio worklets here: https://demo.alphatab.net/crm/audioworklets/ this version has not any improvement beside the buffer fixes and the audio worklets. Most other improvements didn't anyhow have an effect on the load of the UI layer so I rather went back to improving the audio playback only. 

I still plan to bring in some performance improvements separately, but only if they really make sense from a performance and maintainability perspective. 

I see it as most important to fix the crackling audio. Reducing the load on devices in general is a second priority.

**Reply by Reiji777** [reply]
@Danielku15 Thanks for the updates Daniel. Below is the profiling result.
Overall I think it works great. Some issues I encountered below might be edge cases so just to share for your information.
https://drive.google.com/file/d/1wPAR9b948E-IylF0xTrkscDPVoO2X1v3/view?usp=sharing

There are three folders inside:
- Cracklings: when I opened the link for the first time, I heard the cracklings. This couldn't be reproduced after I refreshed the page. I doubt if this was because for the first time, the page was actually using some previously cached script, so just to share as a reference. I noticed in the profiling snapshot there were two workers running, while the other cases only had one.

- No Crackings: After refreshing the page as mentioned above, I don't hear the cracklings any more.

- Cracklings with buzz: The two profiles above are conducted on Chrome, while this one was on the system's default browser. (com.sec.android.app.sbrowser (87.0.4280.141)) I could always hear the cracklings with the buzz noises.

**Comment by lannka** [comment]
FYI that I experienced similar issue on my [OnePlus 7 Pro](https://www.gsmarena.com/oneplus_7_pro-9689.php) phone. Cracks are heard every 5s. Disabling cursor does reduce the amount of cracks.

**Comment by Danielku15** [comment, marked as answer]
I integrated today the improvements and various improvements and fixes are now available via 1.3.0-alpha.127 It combines various fixes which should solve most of the performance issues and improves the playback via audio worklets. Changes are roughly:

1. Added Audio Worklets for playback.
2. Improved cursor positioning and animation by using transforms instead of positions
3. Improved lookup of currently played beat during playback 
4. Added setting which allows disabling of beat cursor animations and element highlighting
5. Use default browser smooth scrolling over custom scrolling animation
6. Some smaller perf improvements. 

https://demo.alphatab.net/crm/crackling/ is running using the updated build if you like to do some further testing.

**Reply by Reiji777** [reply]
@Danielku15 Thanks for the updates! As I was testing the latest version, it seems to be working fine on every device except for iPhone. Just submitted an issue: https://github.com/CoderLine/alphaTab/issues/646
